Fixing SaferCPP warnings in InjectedBundle.cpp

kmonsen opened the pull request:
#### 2394366f6484884f885fb67c76f011821b412a71
<pre>
Fixing SaferCPP warnings in InjectedBundle.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=290078">https://bugs.webkit.org/show_bug.cgi?id=290078</a>
<a href="https://rdar.apple.com/problem/147455928">rdar://problem/147455928</a>

Reviewed by Chris Dumez.

* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundle.cpp:
(WebKit::InjectedBundle::postMessage):
(WebKit::InjectedBundle::postSynchronousMessage):
(WebKit::InjectedBundle::addOriginAccessAllowListEntry):
(WebKit::InjectedBundle::removeOriginAccessAllowListEntry):
(WebKit::InjectedBundle::resetOriginAccessAllowLists):
(WebKit::InjectedBundle::numberOfPages):
(WebKit::InjectedBundle::pageNumberForElementById):
(WebKit::InjectedBundle::pageSizeAndMarginsInPixels):
(WebKit::InjectedBundle::isPageBoxVisible):
(WebKit::InjectedBundle::removeAllWebNotificationPermissions):
(WebKit::InjectedBundle::webNotificationID):
(WebKit::InjectedBundle::liveDocumentURLs):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::protectedNotificationPermissionRequestManager):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/292544@main">https://commits.webkit.org/292544@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c5b4d55cc2bd6cb41e0436404708035cb2915e33

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96305 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15919 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5890 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101372 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46824 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16215 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24352 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73401 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30630 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99308 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12164 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86992 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53738 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11920 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46151 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82033 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4857 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103400 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23372 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17013 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82443 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23623 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83017 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81820 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20545 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26456 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3890 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/16746 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23335 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28490 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22994 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26474 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24735 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/InjectedBundle/InjectedBundle.cpp:
```diff
@@ -167,13 +167,13 @@ void InjectedBundle::setAsynchronousSpellCheckingEnabled(bool enabled)
 int InjectedBundle::numberOfPages(WebFrame* frame, double pageWidthInPixels, double pageHeightInPixels)
 {
-    auto* coreFrame = frame ? frame->coreLocalFrame() : nullptr;
+    auto coreFrame = frame ? frame->protectedCoreLocalFrame() : nullptr;
```
We normally use:
```
RefPtr coreFrame = frame ? frame->coreLocalFrame() : nullptr;
```

Per the safer cpp guidelines, we use `protectedFoo()` functions to avoid local variables, not to initialize them. We also avoid hiding smart pointers behind `auto` in new code to make it clearer when code is safe.

kmonsen replied:
Thank you for explaining, done.

cdumez commented on Source/WebKit/WebProcess/InjectedBundle/InjectedBundle.cpp:
```diff
@@ -184,34 +184,34 @@ int InjectedBundle::pageNumberForElementById(WebFrame* frame, const String& id,
 String InjectedBundle::pageSizeAndMarginsInPixels(WebFrame* frame, int pageIndex, int width, int height, int marginTop, int marginRight, int marginBottom, int marginLeft)
 {
-    auto* coreFrame = frame ? frame->coreLocalFrame() : nullptr;
+    auto coreFrame = frame ? frame->protectedCoreLocalFrame() : nullptr;
```
Same comment as above.

kmonsen replied:
Done

cdumez commented on Source/WebKit/WebProcess/InjectedBundle/InjectedBundle.cpp:
```diff
@@ -184,34 +184,34 @@ int InjectedBundle::pageNumberForElementById(WebFrame* frame, const String& id,
 bool InjectedBundle::isPageBoxVisible(WebFrame* frame, int pageIndex)
 {
-    auto* coreFrame = frame ? frame->coreLocalFrame() : nullptr;
+    auto coreFrame = frame ? frame->protectedCoreLocalFrame() : nullptr;
```
Same as above.

kmonsen replied:
Done

cdumez commented on Source/WebKit/WebProcess/InjectedBundle/InjectedBundle.cpp:
```diff
@@ -314,11 +313,11 @@ InjectedBundle::DocumentIDToURLMap InjectedBundle::liveDocumentURLs(bool exclude
         Page::forEachPage([&](Page& page) {
-            for (auto* frame = &page.mainFrame(); frame; frame = frame->tree().traverseNext()) {
-                auto* localFrame = dynamicDowncast<LocalFrame>(frame);
+            for (RefPtr frame = page.protectedMainFrame(); frame; frame = frame->tree().traverseNext()) {
```
This should remain `&page.mainFrame()`

kmonsen replied:
Done

cdumez commented on Source/WebKit/WebProcess/InjectedBundle/InjectedBundle.cpp:
```diff
@@ -314,11 +313,11 @@ InjectedBundle::DocumentIDToURLMap InjectedBundle::liveDocumentURLs(bool exclude
                 if (!localFrame)
                     continue;
-                auto* document = localFrame->document();
+                auto document = localFrame->protectedDocument();
```
`RefPtr document = localFrame->document();`

kmonsen replied:
Done

cdumez commented on Source/WebKit/WebProcess/InjectedBundle/InjectedBundle.cpp:
```diff
@@ -276,7 +276,7 @@ void InjectedBundle::setUserStyleSheetLocation(const String& location)
 {
 #if ENABLE(NOTIFICATIONS)
-    page->notificationPermissionRequestManager()->removeAllPermissionsForTesting();
+    RefPtr { page }->notificationPermissionRequestManager()->removeAllPermissionsForTesting();
```
This is not right, `page` is safe here since this is a parameter and parameters are protected at the call sites.

The static analysis is about `notificationPermissionRequestManager()`. We should introduce a `WebPage::protectedNotificationPermissionRequestManager()` and call it.

cdumez commented on Source/WebKit/WebProcess/InjectedBundle/InjectedBundle.cpp:
```diff
@@ -285,10 +285,9 @@ void InjectedBundle::removeAllWebNotificationPermissions(WebPage* page)
-    if (!notification)
-        return std::nullopt;
-    return notification->identifier();
+    if (auto* wrapper = jsDynamicCast<JSNotification*>(toJS(toJS(jsContext), jsNotification)))
```
I don't quite understand this change. What was wrong with the original code? Could we have replaced `WebCore::Notification*` with `RefPtr`?

kmonsen replied:
That works, I left it `RefPtr<WebCore::Notification>` as that is more descriptive, or do you prefer just `RefPtr`?

I am sure I tried that before going down that whole path, but it works so now I am not sure.

cdumez replied:
Yes, we prefer to use just `Ref` / `RefPtr` without `<T>`

cdumez reviewed (approved):
LGTM

kmonsen committed:
Fixing SaferCPP warnings in InjectedBundle.cpp
https://bugs.webkit.org/show_bug.cgi?id=290078
rdar://problem/147455928

Reviewed by Chris Dumez.

* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundle.cpp:
(WebKit::InjectedBundle::postMessage):
(WebKit::InjectedBundle::postSynchronousMessage):
(WebKit::InjectedBundle::addOriginAccessAllowListEntry):
(WebKit::InjectedBundle::removeOriginAccessAllowListEntry):
(WebKit::InjectedBundle::resetOriginAccessAllowLists):
(WebKit::InjectedBundle::numberOfPages):
(WebKit::InjectedBundle::pageNumberForElementById):
(WebKit::InjectedBundle::pageSizeAndMarginsInPixels):
(WebKit::InjectedBundle::isPageBoxVisible):
(WebKit::InjectedBundle::removeAllWebNotificationPermissions):
(WebKit::InjectedBundle::webNotificationID):
(WebKit::InjectedBundle::liveDocumentURLs):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::protectedNotificationPermissionRequestManager):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/292544@main
